File: rust/src/attribution/convert.rs
use std::str::FromStr;

use chrono::{Duration, SecondsFormat};
use rust_decimal::prelude::ToPrimitive;
use rust_decimal::Decimal;
use serde_json::Value;

use crate::core::models::{CostConfidence, CostEvent, EventType, PricingSource, Task, TaskStatus};

use super::types::*;
use super::validate::validate_attribution_event_v2;

const GIB: i64 = 1024 * 1024 * 1024;

pub fn to_attribution_event_v2(event: &CostEvent) -> Option<AttributionEventV2> {
    let (component, mut usage, duration) = component_and_usage(event)?;
    if usage.is_empty() {
        append_usage(
            &mut usage,
            AttributionUsageMetric::RequestCount,
            Decimal::ONE,
        );
    }

    let occurred_at = canonical_time(event.occurred_at);
    let has_time_usage = usage.iter().any(|line| {
        matches!(
            line.unit,
            AttributionUsageUnit::Seconds
                | AttributionUsageUnit::VcpuSeconds
                | AttributionUsageUnit::GibSeconds
                | AttributionUsageUnit::GpuSeconds
        )
    });
    let usage_period = if has_time_usage || duration > Decimal::ZERO {
        let micros = duration
            .max(Decimal::ZERO)
            .checked_mul(Decimal::from(1_000_000_i64))
            .and_then(|value| value.round().to_i64())
            .unwrap_or(0);
        Some(AttributionUsagePeriodV2 {
            start_at: canonical_time(event.occurred_at - Duration::microseconds(micros)),
            end_at: Some(occurred_at.clone()),
        })
    } else {
        None
    };

    let converted = AttributionEventV2 {
        schema_version: "2".to_string(),
        event_id: event.event_id.clone(),
        task_id: event.task_id.clone(),
        occurred_at: occurred_at.clone(),
        observed_at: occurred_at,
        component,
        provider: provider_for(event),
        resource: resource_for(event),
        lifecycle: AttributionLifecycleV2 {
            state: AttributionLifecycleState::Final,
            revision: 1,
        },
        usage_period,
        usage,
        cost_evidence: cost_evidence_for(event),
        retry_of: if event.is_retry {
            event.retry_of.clone()
        } else {
            None
        },
    };

    let value = serde_json::to_value(&converted).ok()?;
    let validation = validate_attribution_event_v2(&value);
    if validation.success {
        Some(converted)
    } else {
        let paths: Vec<&str> = validation
            .issues
            .iter()
            .map(|issue| issue.path.as_str())
            .collect();
        eprintln!(
            "[dexcost] event {} cannot be represented by attribution v2: {}",
            event.event_id,
            paths.join(", ")
        );
        None
    }
}

pub fn to_attribution_task_ingest_v1(task: &Task) -> AttributionTaskIngestV1 {
    AttributionTaskIngestV1 {
        task_id: task.task_id.clone(),
        task_type: task.task_type.clone(),
        status: match task.status {
            TaskStatus::Running => "running",
            TaskStatus::Pending => "pending",
            TaskStatus::Success => "success",
            TaskStatus::Failed => "failed",
        }
        .to_string(),
        started_at: canonical_time(task.started_at),
        ended_at: task.ended_at.map(canonical_time),
        metadata: task.metadata.clone(),
        customer_id: task.customer_id.clone(),
        project_id: task.project_id.clone(),
        parent_task_id: task.parent_task_id.clone(),
        experiment_id: task.experiment_id.clone(),
        variant: task.variant.clone(),
        schema_version: "1".to_string(),
    }
}

fn component_and_usage(
    event: &CostEvent,
) -> Option<(AttributionComponent, Vec<AttributionUsageLineV2>, Decimal)> {
    let mut usage = Vec::new();
    match event.event_type {
        EventType::RetryMarker | EventType::GpuUtilizationSignal => None,
        EventType::LlmCall => {
            let cached = Decimal::from(event.cached_tokens.unwrap_or(0).max(0));
            let mut input = Decimal::from(event.input_tokens.unwrap_or(0).max(0));
            let provider = event
                .provider
                .as_deref()
                .unwrap_or_default()
                .to_ascii_lowercase();
            if !provider.contains("anthropic") && !provider.contains("bedrock") && provider != "aws"
            {
                input = (input - cached).max(Decimal::ZERO);
            }
            let cache_write = decimal_detail(&event.details, &["cache_creation_input_tokens"])
                .unwrap_or(Decimal::ZERO);
            let reasoning = decimal_detail(
                &event.details,
                &["reasoning_output_tokens", "reasoning_tokens"],
            )
            .unwrap_or(Decimal::ZERO);
            let output = (Decimal::from(event.output_tokens.unwrap_or(0).max(0)) - reasoning)
                .max(Decimal::ZERO);
            append_usage(&mut usage, AttributionUsageMetric::InputTokens, input);
            append_usage(
                &mut usage,
                AttributionUsageMetric::CacheReadInputTokens,
                cached,
            );
            append_usage(
                &mut usage,
                AttributionUsageMetric::CacheWriteInputTokens,
                cache_write,
            );
            append_usage(&mut usage, AttributionUsageMetric::OutputTokens, output);
            append_usage(
                &mut usage,
                AttributionUsageMetric::ReasoningOutputTokens,
                reasoning,
            );
            Some((AttributionComponent::Llm, usage, Decimal::ZERO))
        }
        EventType::ComputeCost => {
            let duration = decimal_detail(&event.details, &["duration_ms"])
                .unwrap_or(Decimal::ZERO)
                / Decimal::from(1000_i64);
            let duration = if duration.is_zero() {
                decimal_detail(&event.details, &["wall_clock_seconds"]).unwrap_or(Decimal::ZERO)
            } else {
                duration
            };
            append_usage(&mut usage, AttributionUsageMetric::ComputeSeconds, duration);
            append_usage(
                &mut usage,
                AttributionUsageMetric::VcpuSeconds,
                decimal_detail(&event.details, &["vcpu_seconds_used"]).unwrap_or(Decimal::ZERO),
            );
            if let Some(memory) =
                decimal_detail(&event.details, &["memory_bytes_limit", "memory_bytes_peak"])
            {
                append_usage(
                    &mut usage,
                    AttributionUsageMetric::MemoryGibSeconds,
                    memory / Decimal::from(GIB) * duration,
                );
            }
            append_usage(
                &mut usage,
                AttributionUsageMetric::RequestCount,
                decimal_detail(&event.details, &["invocation_count"]).unwrap_or(Decimal::ZERO),
            );
            Some((AttributionComponent::Compute, usage, duration))
        }
        EventType::GpuCost => {
            let duration = decimal_detail(&event.details, &["duration_ms"])
                .unwrap_or(Decimal::ZERO)
                / Decimal::from(1000_i64);
            let measured = decimal_detail(&event.details, &["gpu_seconds_used"]);
            let count = decimal_detail(&event.details, &["gpu_count"]).unwrap_or(Decimal::ONE);
            let mut billed = duration * count;
            if (string_detail(&event.details, &["billing_model"]) == Some("per_gpu_second_active")
                || billed.is_zero())
                && measured.is_some()
            {
                billed = measured.unwrap_or(Decimal::ZERO);
            }
            append_usage(&mut usage, AttributionUsageMetric::GpuSeconds, billed);
            Some((AttributionComponent::Gpu, usage, duration))
        }
        EventType::Network => {
            append_usage(
                &mut usage,
                AttributionUsageMetric::BytesOut,
                decimal_detail(&event.details, &["request_bytes"]).unwrap_or(Decimal::ZERO),
            );
            append_usage(
                &mut usage,
                AttributionUsageMetric::BytesIn,
                decimal_detail(&event.details, &["response_bytes"]).unwrap_or(Decimal::ZERO),
            );
            Some((AttributionComponent::Network, usage, Decimal::ZERO))
        }
        EventType::ExternalCost => {
            let quantity = decimal_detail(&event.details, &["attribution_usage_quantity", "units"])
                .unwrap_or(Decimal::ONE);
            let metric = string_detail(&event.details, &["attribution_usage_metric"])
                .and_then(metric_from_str)
                .unwrap_or_else(|| {
                    metric_for_per(
                        string_detail(&event.details, &["attribution_usage_per"])
                            .unwrap_or("request"),
                    )
                });
            append_usage(&mut usage, metric, quantity);
            Some((AttributionComponent::External, usage, Decimal::ZERO))
        }
    }
}

fn provider_for(event: &CostEvent) -> AttributionProviderIdentityV2 {
    let raw_provider = event.provider.as_deref().unwrap_or_default();
    let raw_lower = raw_provider.to_ascii_lowercase();
    let (mut name, mut service) = if raw_lower.contains("openai") {
        ("openai".to_string(), "responses".to_string())
    } else if raw_lower.contains("anthropic") {
        ("anthropic".to_string(), "messages".to_string())
    } else if raw_lower.contains("bedrock") {
        ("aws".to_string(), "bedrock".to_string())
    } else if raw_lower.contains("gemini") || raw_lower == "google" {
        ("google".to_string(), "generate_content".to_string())
    } else if raw_lower.contains("cohere") {
        ("cohere".to_string(), "chat".to_string())
    } else if raw_lower.contains("vercel") {
        ("vercel".to_string(), "ai_sdk".to_string())
    } else if raw_lower.contains("langchain") {
        ("langchain".to_string(), "chat".to_string())
    } else {
        (canonical_name(raw_provider, "unknown"), "api".to_string())
    };

    if event.event_type != EventType::LlmCall {
        let billing = string_detail(&event.details, &["billing_model"]).unwrap_or_default();
        match event.event_type {
            EventType::ComputeCost => {
                name = if billing.starts_with("azure") {
                    "azure".to_string()
                } else if billing == "gce"
                    || billing == "cloud_functions"
                    || billing.starts_with("cloud_")
                {
                    "google_cloud".to_string()
                } else if billing == "vercel_fluid" {
                    "vercel".to_string()
                } else if billing == "k8s_pod" {
                    "kubernetes".to_string()
                } else if matches!(billing, "lambda" | "fargate" | "ec2") {
                    "aws".to_string()
                } else {
                    canonical_name(raw_provider, "runtime")
                };
                service = canonical_name(
                    first_non_empty(&[Some(billing), event.service_name.as_deref()]),
                    "compute",
                );
            }
            EventType::GpuCost => {
                name = canonical_name(
                    first_non_empty(&[
                        string_detail(&event.details, &["cloud_provider"]),
                        event.provider.as_deref(),
                    ]),
                    "runtime",
                );
                service = canonical_name(billing, "gpu");
            }
            EventType::Network => {
                name = canonical_name(
                    first_non_empty(&[
                        string_detail(&event.details, &["cloud_provider"]),
                        event.provider.as_deref(),
                    ]),
                    "internet",
                );
                service = "egress".to_string();
            }
            EventType::ExternalCost => {
                let external_service = event.service_name.as_deref().unwrap_or("external");
                if let Some(tool) = external_service.strip_prefix("mcp:") {
                    name = "mcp".to_string();
                    service = canonical_name(tool, "tool");
                } else if external_service.contains('.') {
                    name = canonical_name(external_service, "external");
                    service = "http_api".to_string();
                } else {
                    name =
                        canonical_name(raw_provider, &canonical_name(external_service, "external"));
                    service = canonical_name(external_service, "api");
                }
            }
            _ => {}
        }
    }

    AttributionProviderIdentityV2 {
        name,
        service,
        record_id: string_detail(
            &event.details,
            &["provider_record_id", "request_id", "call_sid"],
        )
        .filter(|value| value.len() <= 256)
        .map(str::to_string),
        region: string_detail(&event.details, &["region", "cloud_region"])
            .map(|region| canonical_name(region, "unknown")),
    }
}

fn resource_for(event: &CostEvent) -> Option<AttributionResourceV2> {
    if let Some(model) = event.model.as_deref().filter(|value| !value.is_empty()) {
        return Some(AttributionResourceV2 {
            resource_type: AttributionResourceType::Model,
            id: truncate(model, 256),
        });
    }
    match event.event_type {
        EventType::GpuCost => {
            string_detail(&event.details, &["gpu_sku", "instance_type"]).map(|id| {
                AttributionResourceV2 {
                    resource_type: AttributionResourceType::Sku,
                    id: truncate(id, 256),
                }
            })
        }
        EventType::ComputeCost => string_detail(&event.details, &["instance_type", "architecture"])
            .map(|id| AttributionResourceV2 {
                resource_type: AttributionResourceType::Instance,
                id: truncate(id, 256),
            }),
        _ => None,
    }
}

fn cost_evidence_for(event: &CostEvent) -> Option<AttributionCostEvidenceV2> {
    let amount = positive_quantity(event.cost_usd)?;
    let confidence = confidence_for(&event.cost_confidence);
    match event.pricing_source.as_ref()? {
        PricingSource::ProviderResponse => Some(AttributionCostEvidenceV2 {
            amount,
            currency: "USD".to_string(),
            source: AttributionCostEvidenceSource::ProviderReported,
            confidence: if event.cost_confidence == CostConfidence::Exact {
                AttributionCostConfidence::Exact
            } else {
                AttributionCostConfidence::Estimated
            },
            pricing_version: None,
        }),
        PricingSource::Manual | PricingSource::Custom | PricingSource::UserOverride => {
            Some(AttributionCostEvidenceV2 {
                amount,
                currency: "USD".to_string(),
                source: AttributionCostEvidenceSource::Manual,
                confidence,
                pricing_version: None,
            })
        }
        PricingSource::RateRegistry => sdk_evidence(
            event,
            amount,
            AttributionCostEvidenceSource::SdkRateRegistry,
            confidence,
        ),
        PricingSource::ServiceCatalog | PricingSource::Litellm | PricingSource::Tokencost => {
            sdk_evidence(
                event,
                amount,
                AttributionCostEvidenceSource::SdkCatalog,
                confidence,
            )
        }
        PricingSource::Unknown => None,
    }
}

fn sdk_evidence(
    event: &CostEvent,
    amount: String,
    source: AttributionCostEvidenceSource,
    confidence: AttributionCostConfidence,
) -> Option<AttributionCostEvidenceV2> {
    let pricing_version = event
        .pricing_version
        .clone()
        .or_else(|| string_detail(&event.details, &["pricing_version"]).map(str::to_string))?;
    Some(AttributionCostEvidenceV2 {
        amount,
        currency: "USD".to_string(),
        source,
        confidence: if confidence == AttributionCostConfidence::Exact {
            AttributionCostConfidence::Computed
        } else {
            confidence
        },
        pricing_version: Some(pricing_version),
    })
}

fn confidence_for(confidence: &CostConfidence) -> AttributionCostConfidence {
    match confidence {
        CostConfidence::Exact => AttributionCostConfidence::Exact,
        CostConfidence::Computed => AttributionCostConfidence::Computed,
        CostConfidence::Estimated => AttributionCostConfidence::Estimated,
        CostConfidence::Unknown => AttributionCostConfidence::Unknown,
    }
}

fn append_usage(
    usage: &mut Vec<AttributionUsageLineV2>,
    metric: AttributionUsageMetric,
    quantity: Decimal,
) {
    if let Some(quantity) = positive_quantity(quantity) {
        usage.push(AttributionUsageLineV2 {
            metric,
            quantity,
            unit: metric.canonical_unit(),
        });
    }
}

fn positive_quantity(mut value: Decimal) -> Option<String> {
    if value <= Decimal::ZERO {
        return None;
    }
    if value.scale() > 12 {
        value = value.round_dp(12);
    }
    if value <= Decimal::ZERO {
        return None;
    }
    Some(value.normalize().to_string())
}

fn decimal_detail(
    details: &std::collections::HashMap<String, Value>,
    keys: &[&str],
) -> Option<Decimal> {
    for key in keys {
        let Some(value) = details.get(*key) else {
            continue;
        };
        let raw = match value {
            Value::Number(number) => number.to_string(),
            Value::String(value) => value.trim().to_string(),
            _ => continue,
        };
        if let Ok(parsed) = Decimal::from_str(&raw) {
            return Some(parsed);
        }
    }
    None
}

fn string_detail<'a>(
    details: &'a std::collections::HashMap<String, Value>,
    keys: &[&str],
) -> Option<&'a str> {
    keys.iter().find_map(|key| {
        details
            .get(*key)
            .and_then(Value::as_str)
            .map(str::trim)
            .filter(|value| !value.is_empty())
    })
}

fn metric_from_str(metric: &str) -> Option<AttributionUsageMetric> {
    match metric {
        "input_tokens" => Some(AttributionUsageMetric::InputTokens),
        "output_tokens" => Some(AttributionUsageMetric::OutputTokens),
        "cache_read_input_tokens" => Some(AttributionUsageMetric::CacheReadInputTokens),
        "cache_write_input_tokens" => Some(AttributionUsageMetric::CacheWriteInputTokens),
        "reasoning_output_tokens" => Some(AttributionUsageMetric::ReasoningOutputTokens),
        "characters" => Some(AttributionUsageMetric::Characters),
        "audio_seconds" => Some(AttributionUsageMetric::AudioSeconds),
        "connected_seconds" => Some(AttributionUsageMetric::ConnectedSeconds),
        "recording_seconds" => Some(AttributionUsageMetric::RecordingSeconds),
        "agent_seconds" => Some(AttributionUsageMetric::AgentSeconds),
        "compute_seconds" => Some(AttributionUsageMetric::ComputeSeconds),
        "vcpu_seconds" => Some(AttributionUsageMetric::VcpuSeconds),
        "memory_gib_seconds" => Some(AttributionUsageMetric::MemoryGibSeconds),
        "gpu_seconds" => Some(AttributionUsageMetric::GpuSeconds),
        "request_count" => Some(AttributionUsageMetric::RequestCount),
        "call_count" => Some(AttributionUsageMetric::CallCount),
        "bytes_in" => Some(AttributionUsageMetric::BytesIn),
        "bytes_out" => Some(AttributionUsageMetric::BytesOut),
        "image_count" => Some(AttributionUsageMetric::ImageCount),
        "page_count" => Some(AttributionUsageMetric::PageCount),
        "credit_count" => Some(AttributionUsageMetric::CreditCount),
        _ => None,
    }
}

fn metric_for_per(per: &str) -> AttributionUsageMetric {
    let per = per.to_ascii_lowercase();
    if per.contains("page") {
        AttributionUsageMetric::PageCount
    } else if per.contains("credit") || per.contains("unit") {
        AttributionUsageMetric::CreditCount
    } else if per.contains("image") {
        AttributionUsageMetric::ImageCount
    } else if per.contains("call") || per.contains("sms") || per.contains("message") {
        AttributionUsageMetric::CallCount
    } else if per.contains("character") {
        AttributionUsageMetric::Characters
    } else {
        AttributionUsageMetric::RequestCount
    }
}

fn canonical_name(value: &str, fallback: &str) -> String {
    let lower = value.trim().to_ascii_lowercase();
    let lower = lower
        .strip_prefix("https://")
        .or_else(|| lower.strip_prefix("http://"))
        .unwrap_or(&lower);
    let mut result = String::new();
    let mut replaced = false;
    for ch in lower.chars() {
        if ch.is_ascii_lowercase() || ch.is_ascii_digit() || matches!(ch, '.' | '_' | '-') {
            result.push(ch);
            replaced = false;
        } else if !replaced {
            result.push('_');
            replaced = true;
        }
        if result.len() >= 128 {
            break;
        }
    }
    let normalized = result.trim_matches(['_', '-', '.']);
    if normalized.is_empty() {
        fallback.to_string()
    } else {
        normalized.to_string()
    }
}

fn first_non_empty<'a>(values: &[Option<&'a str>]) -> &'a str {
    values
        .iter()
        .flatten()
        .copied()
        .find(|value| !value.is_empty())
        .unwrap_or("")
}

fn truncate(value: &str, max: usize) -> String {
    value.chars().take(max).collect()
}

fn canonical_time(value: chrono::DateTime<chrono::Utc>) -> String {
    value.to_rfc3339_opts(SecondsFormat::Micros, true)
}

File: typescript/src/attribution/convert.ts
import {
  Decimal,
  canonicalDecimal,
  isoCanonical,
  type CostEvent,
  type Task,
} from "../core/models.js";
import {
  ATTRIBUTION_UNIT_BY_METRIC,
  type AttributionComponent,
  type AttributionCostEvidenceV2,
  type AttributionEventV2,
  type AttributionProviderIdentityV2,
  type AttributionResourceV2,
  type AttributionTaskIngestV1,
  type AttributionUsageLineV2,
  type AttributionUsageMetric,
} from "./types.js";
import { validateAttributionEventV2 } from "./validate.js";

const GIB = new Decimal(1024).pow(3);

function numberDetail(details: Record<string, unknown>, ...keys: string[]): number | undefined {
  for (const key of keys) {
    const value = details[key];
    if (typeof value === "number" && Number.isFinite(value)) return value;
    if (typeof value === "string" && value.trim() !== "") {
      const parsed = Number(value);
      if (Number.isFinite(parsed)) return parsed;
    }
  }
  return undefined;
}

function stringDetail(details: Record<string, unknown>, ...keys: string[]): string | undefined {
  for (const key of keys) {
    const value = details[key];
    if (typeof value === "string" && value.trim() !== "") return value;
  }
  return undefined;
}

function canonicalName(value: string | undefined, fallback: string): string {
  const normalized = (value ?? "")
    .trim()
    .toLowerCase()
    .replace(/^https?:\/\//, "")
    .replace(/[^a-z0-9._-]+/g, "_")
    .replace(/^[_\-.]+|[_\-.]+$/g, "")
    .slice(0, 128);
  return normalized || fallback;
}

function positiveQuantity(value: number | string | Decimal | undefined): string | undefined {
  if (value === undefined) return undefined;
  try {
    const decimal = value instanceof Decimal ? value : new Decimal(String(value));
    if (!decimal.isFinite() || !decimal.gt(0)) return undefined;
    const rounded = decimal.toDecimalPlaces(12);
    if (!rounded.gt(0)) return undefined;
    return canonicalDecimal(rounded);
  } catch {
    return undefined;
  }
}

function usageLine(metric: AttributionUsageMetric, quantity: number | string | Decimal | undefined): AttributionUsageLineV2 | undefined {
  const normalized = positiveQuantity(quantity);
  return normalized === undefined
    ? undefined
    : { metric, quantity: normalized, unit: ATTRIBUTION_UNIT_BY_METRIC[metric] };
}

function compactUsage(lines: Array<AttributionUsageLineV2 | undefined>): AttributionUsageLineV2[] {
  return lines.filter((line): line is AttributionUsageLineV2 => line !== undefined);
}

function providerFor(event: CostEvent): AttributionProviderIdentityV2 {
  const raw = (event.provider ?? "").toLowerCase();
  let name = canonicalName(event.provider, "unknown");
  let service = "api";
  if (raw.includes("openai")) [name, service] = ["openai", "responses"];
  else if (raw.includes("anthropic")) [name, service] = ["anthropic", "messages"];
  else if (raw.includes("bedrock")) [name, service] = ["aws", "bedrock"];
  else if (raw.includes("gemini") || raw === "google") [name, service] = ["google", "generate_content"];
  else if (raw.includes("cohere")) [name, service] = ["cohere", "chat"];
  else if (raw.includes("vercel")) [name, service] = ["vercel", "ai_sdk"];
  else if (raw.includes("langchain")) [name, service] = ["langchain", "chat"];

  if (event.eventType !== "llm_call") {
    const billingModel = stringDetail(event.details, "billing_model");
    const serviceName = event.serviceName;
    if (event.eventType === "compute_cost") {
      if (billingModel?.startsWith("azure")) name = "azure";
      else if (billingModel === "gce" || billingModel?.startsWith("cloud_") || billingModel === "cloud_functions") name = "google_cloud";
      else if (billingModel === "vercel_fluid") name = "vercel";
      else if (billingModel === "k8s_pod") name = "kubernetes";
      else if (billingModel === "lambda" || billingModel === "fargate" || billingModel === "ec2") name = "aws";
      else name = canonicalName(event.provider, "runtime");
      service = canonicalName(billingModel ?? serviceName, "compute");
    } else if (event.eventType === "gpu_cost") {
      name = canonicalName(stringDetail(event.details, "cloud_provider") ?? event.provider, "runtime");
      service = canonicalName(billingModel, "gpu");
    } else if (event.eventType === "network") {
      name = canonicalName(stringDetail(event.details, "cloud_provider") ?? event.provider, "internet");
      service = "egress";
    } else if (event.eventType === "retry_marker") {
      name = "dexcost";
      service = "retry";
    } else {
      const rawService = serviceName ?? "external";
      if (rawService.startsWith("mcp:")) {
        name = "mcp";
        service = canonicalName(rawService.slice(4), "tool");
      } else if (rawService.includes(".")) {
        name = canonicalName(rawService, "external");
        service = "http_api";
      } else {
        name = canonicalName(event.provider, canonicalName(rawService, "external"));
        service = canonicalName(rawService, "api");
      }
    }
  }

  const provider: AttributionProviderIdentityV2 = { name, service };
  const recordId = stringDetail(event.details, "provider_record_id", "request_id", "call_sid");
  const region = stringDetail(event.details, "region", "cloud_region");
  if (recordId !== undefined && recordId.length <= 256) provider.record_id = recordId;
  if (region !== undefined) provider.region = canonicalName(region, "unknown");
  return provider;
}

function resourceFor(event: CostEvent): AttributionResourceV2 | undefined {
  const explicitType = stringDetail(event.details, "attribution_resource_type");
  const explicitId = stringDetail(event.details, "attribution_resource_id");
  if (
    explicitId !== undefined &&
    explicitType !== undefined &&
    ["model", "sku", "instance", "endpoint", "session", "other"].includes(explicitType)
  ) {
    return { type: explicitType as AttributionResourceV2["type"], id: explicitId.slice(0, 256) };
  }
  if (event.model) return { type: "model", id: event.model.slice(0, 256) };
  if (event.eventType === "gpu_cost") {
    const sku = stringDetail(event.details, "gpu_sku", "instance_type");
    if (sku) return { type: "sku", id: sku.slice(0, 256) };
  }
  if (event.eventType === "compute_cost") {
    const instance = stringDetail(event.details, "instance_type", "architecture");
    if (instance) return { type: "instance", id: instance.slice(0, 256) };
  }
  if (event.eventType === "retry_marker") {
    const reason = event.retryReason?.trim();
    if (reason) return { type: "other", id: reason.slice(0, 256) };
  }
  return undefined;
}

function evidenceFor(event: CostEvent): AttributionCostEvidenceV2 | undefined {
  const amount = positiveQuantity(event.costUsd);
  if (amount === undefined) return undefined;
  if (event.eventType === "retry_marker") {
    return { amount, currency: "USD", source: "manual", confidence: "exact" };
  }
  const source = event.pricingSource;
  if (source === "provider_response") {
    return {
      amount,
      currency: "USD",
      source: "provider_reported",
      confidence: event.costConfidence === "exact" ? "exact" : "estimated",
    };
  }
  if (source === "manual" || source === "custom") {
    return { amount, currency: "USD", source: "manual", confidence: event.costConfidence };
  }
  const isSdkCatalog = source === "service_catalog"
    || source === "litellm"
    || source === "tokencost"
    || source?.startsWith("compute_catalog:") === true
    || source?.startsWith("gpu_catalog:") === true
    || source?.startsWith("egress_catalog:") === true;
  const mapped = source === "rate_registry" ? "sdk_rate_registry"
    : isSdkCatalog ? "sdk_catalog"
      : undefined;
  if (mapped === undefined || !event.pricingVersion) return undefined;
  return {
    amount,
    currency: "USD",
    source: mapped,
    confidence: event.costConfidence === "exact" ? "computed" : event.costConfidence,
    pricing_version: event.pricingVersion,
  };
}

function componentAndUsage(event: CostEvent): {
  component: AttributionComponent;
  usage: AttributionUsageLineV2[];
  durationSeconds?: number;
} | null {
  const details = event.details;
  switch (event.eventType) {
    case "gpu_utilization_signal":
      return null;
    case "retry_marker":
      return {
        component: "external",
        usage: [usageLine("request_count", 1)!],
      };
    case "llm_call": {
      const cached = event.cachedTokens ?? 0;
      const provider = (event.provider ?? "").toLowerCase();
      const cacheCountersAreDisjoint = provider.includes("anthropic")
        || provider.includes("bedrock")
        || provider === "aws";
      const input = cacheCountersAreDisjoint
        ? event.inputTokens
        : Math.max(0, (event.inputTokens ?? 0) - cached);
      const cacheWrite = numberDetail(details, "cache_creation_input_tokens");
      const reasoning = numberDetail(details, "reasoning_output_tokens", "reasoning_tokens");
      const output = reasoning === undefined ? event.outputTokens : Math.max(0, (event.outputTokens ?? 0) - reasoning);
      const usage = compactUsage([
        usageLine("input_tokens", input),
        usageLine("cache_read_input_tokens", cached),
        usageLine("cache_write_input_tokens", cacheWrite),
        usageLine("output_tokens", output),
        usageLine("reasoning_output_tokens", reasoning),
      ]);
      if (usage.length === 0) usage.push(usageLine("request_count", 1)!);
      return { component: "llm", usage };
    }
    case "compute_cost": {
      const durationSeconds = (numberDetail(details, "duration_ms") ?? 0) / 1000
        || numberDetail(details, "wall_clock_seconds") || 0;
      const memoryBytes = numberDetail(details, "memory_bytes_limit")
        ?? numberDetail(details, "memory_bytes_peak");
      return {
        component: "compute",
        durationSeconds,
        usage: compactUsage([
          usageLine("compute_seconds", durationSeconds),
          usageLine("vcpu_seconds", numberDetail(details, "vcpu_seconds_used")),
          usageLine("memory_gib_seconds", memoryBytes === undefined ? undefined : new Decimal(memoryBytes).div(GIB).times(durationSeconds)),
          usageLine("request_count", numberDetail(details, "invocation_count")),
        ]),
      };
    }
    case "gpu_cost": {
      const durationSeconds = (numberDetail(details, "duration_ms") ?? 0) / 1000;
      const measured = numberDetail(details, "gpu_seconds_used");
      const gpuCount = numberDetail(details, "gpu_count") ?? 1;
      const billingModel = stringDetail(details, "billing_model") ?? "";
      const billedSeconds = billingModel === "per_gpu_second_active" ? measured : durationSeconds * gpuCount;
      return { component: "gpu", durationSeconds, usage: compactUsage([usageLine("gpu_seconds", billedSeconds ?? measured)]) };
    }
    case "network":
      return {
        component: "network",
        usage: compactUsage([
          usageLine("bytes_out", numberDetail(details, "request_bytes")),
          usageLine("bytes_in", numberDetail(details, "response_bytes")),
        ]),
      };
    case "external_cost": {
      const explicitQuantity = numberDetail(details, "attribution_usage_quantity");
      const explicitMetric = stringDetail(details, "attribution_usage_metric");
      const explicitComponent = stringDetail(details, "attribution_component");
      const per = canonicalName(stringDetail(details, "attribution_usage_per"), "request");
      const inferredMetric: AttributionUsageMetric = per.includes("page") ? "page_count"
        : per.includes("credit") ? "credit_count"
          : per.includes("image") ? "image_count"
            : per.includes("call") ? "call_count"
              : per.includes("character") ? "characters"
                : "request_count";
      const metric = explicitMetric !== undefined && explicitMetric in ATTRIBUTION_UNIT_BY_METRIC
        ? explicitMetric as AttributionUsageMetric
        : inferredMetric;
      const component: AttributionComponent = explicitComponent === "speech_to_text"
        ? "speech_to_text"
        : "external";
      return {
        component,
        durationSeconds: numberDetail(details, "attribution_usage_duration_seconds"),
        usage: compactUsage([usageLine(metric, explicitQuantity ?? 1)]),
      };
    }
  }
}

/** Convert the SDK's durable v1 capture model into the strict v2 wire event. */
export function toAttributionEventV2(event: CostEvent): AttributionEventV2 | null {
  const mapped = componentAndUsage(event);
  if (mapped === null) return null;
  if (mapped.usage.length === 0) mapped.usage.push(usageLine("request_count", 1)!);
  const occurredAt = isoCanonical(event.occurredAt);
  const converted: AttributionEventV2 = {
    schema_version: "2",
    event_id: event.eventId,
    task_id: event.taskId,
    occurred_at: occurredAt,
    // CostEvent predates observed_at. Using occurred_at is stable across
    // retries; generating "now" here would change the ledger payload hash.
    observed_at: occurredAt,
    component: mapped.component,
    provider: providerFor(event),
    lifecycle: { state: "final", revision: 1 },
    usage: mapped.usage,
  };
  const resource = resourceFor(event);
  if (resource !== undefined) converted.resource = resource;
  const evidence = evidenceFor(event);
  if (evidence !== undefined) converted.cost_evidence = evidence;
  if (event.isRetry && event.retryOf) converted.retry_of = event.retryOf;
  const hasTimeBasedUsage = mapped.usage.some((line) => line.unit.endsWith("Seconds"));
  if (hasTimeBasedUsage || (mapped.durationSeconds !== undefined && mapped.durationSeconds > 0)) {
    const startOffsetMs = mapped.durationSeconds !== undefined && mapped.durationSeconds > 0
      ? mapped.durationSeconds * 1000
      : 0;
    converted.usage_period = {
      start_at: isoCanonical(new Date(event.occurredAt.getTime() - startOffsetMs)),
      end_at: occurredAt,
    };
  }

  const validation = validateAttributionEventV2(converted);
  if (!validation.success) {
    console.warn(`[dexcost] Event ${event.eventId} cannot be represented by attribution v2: ${validation.issues.map((issue) => issue.path).join(", ")}`);
    return null;
  }
  return converted;
}

export function toAttributionTaskIngestV1(task: Task): AttributionTaskIngestV1 {
  return {
    task_id: task.taskId,
    task_type: task.taskType,
    status: task.status,
    started_at: isoCanonical(task.startedAt),
    ended_at: task.endedAt ? isoCanonical(task.endedAt) : null,
    metadata: task.metadata,
    customer_id: task.customerId ?? null,
    project_id: task.projectId ?? null,
    parent_task_id: task.parentTaskId ?? null,
    experiment_id: task.experimentId ?? null,
    variant: task.variant ?? null,
    schema_version: "1",
  };
}
